src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,353 @@
+        long movingEndByte = partitionEndByteRange;
+        long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
+        boolean newLineFound = false;
+        while (!newLineFound) {
+          movingEndByte += MyriaConstants.BYTE_OVERLAP_PARALLEL_INGEST;
+          // Create a stream to look for the new line
+          InputStream trailingEndInputStream =
+              source.getInputStream(partitionEndByteRange, movingEndByte);
+          InputStreamReader startStreamReader = new InputStreamReader(trailingEndInputStream);
+          int dataChar = startStreamReader.read();
+          while (dataChar != -1) {
+            char currentChar = (char) dataChar;
+            if (currentChar == '\n' || currentChar == '\r') {
+              newLineFound = true;
+              // Re-initialize the parser with the last row only
+              InputStream beginningOfRecord =
+                  source.getInputStream(bytePositionAtBeginningOfRecord, partitionEndByteRange);
+              InputStream concatenateEndOfRecord =
+                  source.getInputStream(partitionEndByteRange + 1, movingEndByte);
+              partitionInputStream =
+                  new SequenceInputStream(beginningOfRecord, concatenateEndOfRecord);
+              parser =
+                  new CSVParser(
+                      new BufferedReader(new InputStreamReader(partitionInputStream)),
+                      CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
+                      bytePositionAtBeginningOfRecord,
+                      0);
+              iterator = parser.iterator();
+              onLastRow = true;
+              record = iterator.next();
+              break;
+            }
+            dataChar = startStreamReader.read();
+          }
+          startStreamReader.close();
+        }
+      }
+      // This is a partial fragment, columns are missing and we need to read more bytes
+      if (record.size() < schema.numColumns()) {
+        if (lineNumber - 1 != 0 && !isLastWorker) {
+          onLastRow = true;
+          long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
+          if (!fixingStartByte) {
+            fixingStartByte = true;
+            partitionStartByteRange += bytePositionAtBeginningOfRecord;
+          }
+          partitionEndByteRange += byteOverlap;
+          InputStream overlapStream =
+              source.getInputStream(partitionStartByteRange, partitionEndByteRange);
+          partitionInputStream = new SequenceInputStream(partitionInputStream, overlapStream);
+          parser =
+              new CSVParser(
+                  new BufferedReader(new InputStreamReader(partitionInputStream)),
+                  CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape));
+          iterator = parser.iterator();
+          byteOverlap *= 2;
+        } else {
+          discardedRecord = true;
```

Comment:
Again, this will make us discard any partial record that spans the entire original partition (which could easily happen on the first worker).